source/slang/slang-ir-autodiff-primal-hoist.cpp

```diff
@@ -1046,6 +1063,174 @@ void AutodiffCheckpointPolicyBase::collectInductionValues(IRGlobalValueWithCode*
+                else
+                    statement = inverseStatement.negated(); // Use ~B <=> ~breakFlag
+            }
+
+            // We found a relation on the parameter at the loop exit, and we also proved that
+            // if the relation holds, the loop must exit.
+            //
+            // If we have an inequality + information that a value is an inductive (i.e. follows a
+            // sequence of the form `start + i * step`), then we can use that to compute the exact
+            // value at the loop exit.
+            //
+            // We can do this by solving the inequality for the parameter, using the inductive value
+            // as the counter variable.
+            //
+            if (inductionValueInfo->kind == LoopInductionValueInfo::Kind::AffineFunctionOfCounter)
+            {
+                auto counterOffset = getConstantIntegerValue(inductionValueInfo->counterOffset);
+                auto counterFactor = inductionValueInfo->counterFactor;
+
+                SLANG_ASSERT(statement.type == SimpleRelation::Type::IntegerRelation);
+                auto relationValue = statement.integerValue;
+
+                auto recordExitValue = [&](IRIntegerValue exitIValue, IRIntegerValue exitParamValue)
+                {
+                    this->loopExitValueInsts[param] =
+                        builder.getIntValue(param->getDataType(), exitParamValue);
+
+                    // The interesting part is that since we know that this variable is an bijective
+                    // function of the loop counter, we can also compute the loop counter's exit
+                    // value.
+                    //
+                    // Since this can come from multiple parameters, we'll verify to make sure that
+                    // there are no contradictions.
+                    //
+                    IRInst* loopCounterExitValue;
+                    if (this->loopExitValueInsts.tryGetValue(loopCounter, loopCounterExitValue))
+                    {
+                        auto loopCounterExitIValue = getConstantIntegerValue(loopCounterExitValue);
+                        if (loopCounterExitIValue != exitIValue)
+                        {
+                            SLANG_ASSERT(!"contradictory loop exit values");
+                        }
+                    }
+                    else
+                    {
+                        this->loopExitValueInsts[loopCounter] =
+                            builder.getIntValue(loopCounter->getDataType(), exitIValue);
+                    }
+                };
+
+                if (counterFactor > 0 && statement.comparator == SimpleRelation::GreaterThanEqual)
+                {
+                    // Find the smallest value that satisfies counterFactor * i + counterOffset >=
+                    // relationValue
+                    //
+                    IRIntegerValue exitIValue =
```

Comment:
This kind of computation without considering the original type of the induction variable can be dangerous because it doesn't factor in possible int overflow behavior. We need to take extra care to make sure we are handling things like int8_t arithmetic correctly.